sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -116,7 +116,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
         operatorOptimizationRuleSet.filterNot(_ == InferFiltersFromConstraints)
       Batch("Operator Optimization before Inferring Filters", fixedPoint,
         rulesWithoutInferFiltersFromConstraints: _*) ::
-      Batch("Infer Filters", Once,
+      Batch("Infer Filters", fixedPoint,
+        PushDownPredicates,
```

Comment:
My concern is that the change above might cause the issue described in #19149. Instead, (just a suggestion), we couldn't improve  `InferFiltersFromConstraints` to cover the case you pointed out in the PR description?

Reply:
Unfortunately, there is no UT in the commit to see what was the exact issue and why `InferFiltersFromConstraints` needs to be separated entirely from other rules.

If I get the issue described in https://issues.apache.org/jira/browse/SPARK-21652 right, if we combine `InferFiltersFromConstraints` with other rules into a batch then we can end up in an infinite loop in that batch. But this is only because `InferFiltersFromConstraints` can create a new constraint that a subsequent rule removes. The example in https://issues.apache.org/jira/browse/SPARK-21652 required running `ConstantPropagation`, `ConstantFolding` and then `BooleanSimplification` to remove such an inferred constraint/filter.
I think if we combine `InferFiltersFromConstraints` with other rules that doesn't reduce constraints (like `PushDownPredicates`) then we are still good. But I might be wrong so please share your thoughts on this.

Sidebar:
I think the source of the above mentioned loop is that `ConstantPropagation` doesn't propagate constants into a join condition (it handles filters only). I had an old PR to enhance that rule where I also commented why it is not so simple to propagate constants into a join: https://github.com/apache/spark/pull/24553/files#diff-d43484d56a4d9991066b5c00d12ec2465c75131e055fc02ee7fb6dfd45b5006fR76-R79 but it is doable if we fix https://issues.apache.org/jira/browse/SPARK-30598 (https://github.com/apache/spark/pull/27309).
But I'm not saying that `ConstantPropagation` is the only reduction rule that can collide with `InferFiltersFromConstraints`.

Reply:
> I think if we combine InferFiltersFromConstraints with other rules that doesn't reduce constraints (like PushDownPredicates) then we are still good. But I might be wrong so please share your thoughts on this.

Yea, I don't have a specific query to deny your thought and I also think it is true, but, in my current feeling, I a bit hesitate to change `Once` -> `fixedPoint ` for this batch without adding no logic to avoid the infinite loop case (and without any evidence that it will not affect users queries...). WDTY, @cloud-fan @viirya ?

Reply:
I initially looked at trying to see if InferFiltersFromConstraints could itself handle this, but I think I could not figure out a way to solve the case in the unit test. I also don't think that this can cause an infinite loop, but let me take another look at this to see if we can either be sure that it cannot result in an infinite loop or if we can push this into InferFiltersFromConstraints so that we don't need the fixedPoint.

Reply:
Not look at `InferFiltersFromConstraints`, but I tend to agree with @maropu's initial comment. Can we possibly make `InferFiltersFromConstraints` deal with this kind of case?